Tests/Test.cs

```diff
@@ -0,0 +1,327 @@
+using Microsoft.Extensions.Time.Testing;
+
+namespace Tests;
+
+using static ShoppingCartCommand;
+using static ShoppingCartEvent;
+using static ShoppingCart;
+
+// Types Aliases
+using ProductItems = ImmutableDictionary<string, int>;
+using ProductItem = (Guid ProductId, int Quantity);
+using PricedProductItem = (Guid ProductId, int Quantity, decimal Price);
+
+// Commands
+public abstract record ShoppingCartCommand
+{
+    public record AddProduct(
+        ProductItem ProductItem,
+        Guid? ClientId = null
+    ): ShoppingCartCommand;
+
+    public record RemoveProduct(
+        PricedProductItem ProductItem
+    ): ShoppingCartCommand;
+
+    public record Confirm(
+        Guid ClientId
+    ): ShoppingCartCommand;
+
+    public record Cancel: ShoppingCartCommand;
+
+    private ShoppingCartCommand() { }
+}
+
+public abstract record ShoppingCartEvent
+{
+    public record Opened(
+        Guid? ClientId,
+        DateTimeOffset OpenedAt
+    ): ShoppingCartEvent;
+
+    public record ProductAdded(
+        PricedProductItem ProductItem,
+        DateTimeOffset AddedAt
+    ): ShoppingCartEvent;
+
+    public record ProductRemoved(
+        PricedProductItem ProductItem,
+        DateTimeOffset RemovedAt
+    ): ShoppingCartEvent;
+
+    public record Confirmed(
+        Guid ClientId,
+        DateTimeOffset ConfirmedAt
+    ): ShoppingCartEvent;
+
+    public record Cancelled(
+        DateTimeOffset CanceledAt
+    ): ShoppingCartEvent;
+
+    private ShoppingCartEvent() { }
+}
+
+public abstract record ShoppingCart
+{
+    public record Empty: ShoppingCart;
+
+    public record Pending(ProductItems ProductItems): ShoppingCart;
+
+    public record Closed: ShoppingCart;
+
+    private ShoppingCart() { }
+}
+
+// Primary constructor & TimeProvider
+// Then Collection expressions and advanced pattern matching
+public class ShoppingClassDecider(
+    IProductPriceCalculator priceCalculator,
+    TimeProvider timeProvider
+)
+{
+    public ShoppingCartEvent[] Decide(ShoppingCartCommand command, ShoppingCart state) =>
+        (state, command) switch
+        {
+            (Empty, AddProduct (var productItem, var clientId)) =>
+            [
+                new Opened(clientId, Now),
+                new ProductAdded(priceCalculator.Calculate(productItem).Single(), Now)
+            ],
+
+            (Pending, AddProduct (var productItem, _)) =>
+            [
+                new ProductAdded(priceCalculator.Calculate(productItem).Single(), Now)
+            ],
+
+            (Pending pending, RemoveProduct((var productId, var toRemove, var price) pricedProductItem)) =>
+                pending.ProductItems.TryGetValue($"{productId}{price}", out var quantity) && quantity >= toRemove
+                    ?
+                    [
+                        new ProductRemoved(pricedProductItem, Now)
+                    ]
+                    : throw new InvalidOperationException("Not enough Products!"),
```

Comment:
just handle the command idempotently by ignoring
(but really, this should of course be a changequantity so a) the command can be idempotent b) the events and ingesting into a read model can become simpler too)

Reply:
Wondering why not return a DomainEvent instead of an exception?

Reply:
In 99% of cases you want to design commands such that they can be applied idempotently - having a command that's contextual (reduce quantity by 1) is bad news. The command should encode an intent, e.g. remove the item, set the quantity to N. The job of the decision function is to map that intent to an event iff there's something to be done to manifest the need. 

In some cases recording a domain event can make sense if there's something to be gained https://thinkbeforecoding.github.io/FsUno.Prod/Exception%20or%20Event.html - be carefully though - that can be overdone very easily

In this case, if you are removing an item that does not exist in the context of a web app or a mobile app, that's just a bad protocol or programmer error. We are not trying to build a video game where you need to click on a product just right to remove it!

Either way an exception is definitely not the answer (though making it return an Either with a Fail would be worse...)

There are cases where throwing an exception makes sense - normally if there is some invariant that should not have been contravened, a double check can be a good idea from a defence in depth perspective - but only if there is a reasonable way in which an outer layer should have been able to deterministically short circuit sending the command in the first instance. In all other cases, the handler should just do its job of idempotently mappping the intent of the commmand to an event (or none)

Reply:
I'm throwing exceptions where I don't intend anyone to retry the operation. Here, I don't have any retriable scenario. I agree that in the perfect world, we should analyse each scenario and try to make it idempotent on the domain level, but from what I noticed that's actually rarely needed. Too often, that ends up in the analysis-paralysis. I prefer to have explicit results when I know that actually, I'm going to need it. I prefer to keep it simple until a simple solution is good enough from the business perspective.